Add `requireKeySystemAccessOnStart` (fixes Chrome PIPELINE_DECODE_ERROR on clear to encrypted transition)

grabofus opened the pull request:
### This PR will...

Request key system access based on the eme configuration (with `config.requireKeySystemAccessOnStart`  enabled) and block loading clear content until the media keys are set.

This resolves an issue on Chrome where if clear content is encrypted into the SourceBuffer followed by setting media keys then transitioning to encrypted content, Chrome will die with `PIPELINE_DECODE_ERROR`:
![image](https://github.com/user-attachments/assets/fb9c354a-4f44-4fa0-bc4e-49873f2bec7e)

### Why is this Pull Request needed?

- If the HLS stream starts with clear content only in the manifest, and later transitions to encrypted content
- If the HLS stream starts with mixed content in the manifest, but the first buffered segment is unencrypted. This creates a race-condition today, where the first clear segment might load before the media keys could be set. This has been noticed on MediaTailor live stream with prerolls, where the prerolls were loaded from disk cache, loading faster than it took the media keys to be set.

### Are there any points in the code the reviewer needs to double check?

### Resolves issues:

https://github.com/video-dev/hls.js/issues/4230
https://github.com/video-dev/hls.js/issues/5753

### Checklist

- [x] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] API or design changes are documented in API.md

grabofus commented:
Issues reproducible with hls.js 1.6.2 at:
https://sample-videos-zyrkp2nj.s3-eu-west-1.amazonaws.com/test-20250422/master_clear_to_cbcs.html
https://sample-videos-zyrkp2nj.s3-eu-west-1.amazonaws.com/test-20250422/master_clear_to_cbcs_vod.html

Same stream with this fix:
https://sample-videos-zyrkp2nj.s3-eu-west-1.amazonaws.com/test-20250422/master_clear_to_cbcs_preset_media_keys.html
https://sample-videos-zyrkp2nj.s3-eu-west-1.amazonaws.com/test-20250422/master_clear_to_cbcs_preset_media_keys_vod.html

robwalch commented:
@grabofus I'm seeing very mixed results with https://sample-videos-zyrkp2nj.s3-eu-west-1.amazonaws.com/test-20250422/master_clear_to_cbcs_preset_media_keys_vod.html

Occasionally, Chrome successfully plays through and allows seeking between the discontinuities. At other times, it still fails with a pipeline error:

Media error in dev console:
```
$ video.error
MediaError {code: 3, message: 'PIPELINE_ERROR_DECODE: video decoder reinitialization failed'}
```

Media tab logs:
```
Selected VideoToolboxVideoDecoder for video decoding, config: codec: h264, profile: h264 baseline, level: not available, alpha_mode: is_opaque, coded size: [640,360], visible rect: [0,0,640,360], natural size: [640,360], has extra data: false, encryption scheme: Unencrypted, rotation: 0°, flipped: 0, color space: {primaries:BT709, transfer:BT709, matrix:BT709, range:LIMITED}

video decoder config changed midstream, new config: codec: h264, profile: h264 baseline, level: not available, alpha_mode: is_opaque, coded size: [640,360], visible rect: [0,0,640,360], natural size: [640,360], has extra data: false, encryption scheme: CBCS, rotation: 0°, flipped: 0, color space: {primaries:BT709, transfer:BT709, matrix:BT709, range:LIMITED}

Cannot select DecryptingVideoDecoder for video decoding

video decoder reinitialization failed

  Error Group: DecoderStatus
  Error Code: 202
  Stacktrace: media/filters/decoder_selector.cc:213
  
  Error Group: PipelineStatus
  Error Code: 3
  Stacktrace: media/renderers/video_renderer_impl.cc:596
  
  Caused by:
    Error Group: DecoderStatus
    Error Code: 108
    Stacktrace: media/filters/decoder_stream.cc:196
```


On the occasion that I observed it play through without error, it detected the audio decoder config change first:

```
Cannot select DecryptingVideoDecoder for video decoding

Selected VideoToolboxVideoDecoder for video decoding, config: codec: h264, profile: h264 baseline, level: not available, alpha_mode: is_opaque, coded size: [640,360], visible rect: [0,0,640,360], natural size: [640,360], has extra data: false, encryption scheme: Unencrypted, rotation: 0°, flipped: 0, color space: {primaries:BT709, transfer:BT709, matrix:BT709, range:LIMITED}

audio decoder config changed midstream, new config: codec: aac, profile: unknown, bytes_per_channel: 2, channel_layout: STEREO, channels: 2, samples_per_second: 48000, sample_format: Signed 16-bit, bytes_per_frame: 4, seek_preroll: 0us, codec_delay: 0, has extra data: false, encryption scheme: CBCS, discard decoder delay: false, target_output_channel_layout: STEREO, target_output_sample_format: Unknown sample format, has aac extra data: true

Selected DecryptingDemuxerStream for audio decryption, config: codec: aac, profile: unknown, bytes_per_channel: 2, channel_layout: STEREO, channels: 2, samples_per_second: 48000, sample_format: Signed 16-bit, bytes_per_frame: 4, seek_preroll: 0us, codec_delay: 0, has extra data: false, encryption scheme: CBCS, discard decoder delay: false, target_output_channel_layout: STEREO, target_output_sample_format: Unknown sample format, has aac extra data: true

Selected FFmpegAudioDecoder for audio decoding, config: codec: aac, profile: unknown, bytes_per_channel: 2, channel_layout: STEREO, channels: 2, samples_per_second: 48000, sample_format: Signed 16-bit, bytes_per_frame: 4, seek_preroll: 0us, codec_delay: 0, has extra data: false, encryption scheme: Unencrypted, discard decoder delay: false, target_output_channel_layout: STEREO, target_output_sample_format: Unknown sample format, has aac extra data: true

video decoder config changed midstream, new config: codec: h264, profile: h264 baseline, level: not available, alpha_mode: is_opaque, coded size: [640,360], visible rect: [0,0,640,360], natural size: [640,360], has extra data: false, encryption scheme: CBCS, rotation: 0°, flipped: 0, color space: {primaries:BT709, transfer:BT709, matrix:BT709, range:LIMITED}

Selected DecryptingVideoDecoder for video decoding, config: codec: h264, profile: h264 baseline, level: not available, alpha_mode: is_opaque, coded size: [640,360], visible rect: [0,0,640,360], natural size: [640,360], has extra data: false, encryption scheme: CBCS, rotation: 0°, flipped: 0, color space: {primaries:BT709, transfer:BT709, matrix:BT709, range:LIMITED}
```

grabofus commented:
@robwalch this branch was missing an earlier commit from our fork that attempts to set media keys as soon as they are created, instead of waiting for the `encrypted` event. I've added that commit now, and republished the test pages.

https://sample-videos-zyrkp2nj.s3-eu-west-1.amazonaws.com/test-20250422/master_clear_to_cbcs_preset_media_keys.html
https://sample-videos-zyrkp2nj.s3-eu-west-1.amazonaws.com/test-20250422/master_clear_to_cbcs_preset_media_keys_vod.html

robwalch reviewed (changes requested):
Thanks for the update. It works quite nicely. 

Can you look at why this unit test is failing?

https://github.com/video-dev/hls.js/actions/runs/14861488418/job/41727400833?pr=7216

https://github.com/video-dev/hls.js/blob/3ca6f267eeea6ef2697efc90fa9c7a2bbed8dfb5/tests/unit/controller/eme-controller.ts#L106

robwalch commented on src/controller/eme-controller.ts:
```diff
@@ -292,6 +294,7 @@ class EMEController extends Logger implements ComponentAPI {
                   certificate,
                 );
               }
+              this.attemptSetMediaKeys(keySystem, mediaKeys);
```
This change is resulting in a second call to `setMediaKeys` (failing unit test assertion that setMediaKeys is only called once). Is it necessary?

grabofus commented:
@robwalch Thanks for the review! I've incorporated the changes you've suggested and updated the test pages.
All unit tests are passing now. However, I did notice that some logs are duplicated now:
```
[...]
[log] > [eme]: Access for key-system "com.widevine.alpha" obtained
[log] > [eme]: Create media-keys for "com.widevine.alpha"
[log] > [eme]: Media-keys created for "com.widevine.alpha"
[log] > [eme]: Setting media-keys for "com.widevine.alpha"
[log] > [eme]: Setting media-keys for "com.widevine.alpha"
[log] > [eme]: Media-keys set for "com.widevine.alpha"
[log] > [eme]: Media-keys set for "com.widevine.alpha"
[log] > [stream-controller]: FRAG_LOADING->IDLE
[log] > [stream-controller]: Loading main sn: 0 of level 0 (frag:[0.000-6.000]) cc: 0 [0-2], target: 0
[...]
```

robwalch commented:
Hi @grabofus. Thanks for making those changes.

> All unit tests are passing now. However, I did notice that some logs are duplicated now:

Any thoughts on deduping media-keys setting?

> On a separate note, should the requireKeySystemAccessOnStart config option stay disabled by default?

I think it should for v1.6.3. File an issue for defaulting it to enabling it in v1.7.

robwalch commented:
@grabofus I merged and followed up with #7284 to address the issue is mediaKeys being set multiple times. Please let me know if you spot anything else.
